### ISSUE TITLE
subs dict result changed (and inconsistent) in 1.6 with functions

### ISSUE THREAD

**Issue by cmbant** [sympy/sympy#19326]
This code

```
from sympy import Function
from sympy.abc import t

x, y = Function('x')(t), Function('y')(t)
(x*y).subs({x: x + 1, y: x})
```

produces the expected result `(x(t) + 1)*x(t)` with 1.5.1 and earlier, but with 1.6rc1 produces (x(t) + 1)**2. Note that this is even inconsistent with substituting in the ordered dict items order (I believe dict substitution should be independent of order and not make any re-substitutions). I don't see anything in the change log that could justify this behaviour, so assuming a bug (and a tricky one to isolate, took me half a day to figure out why a complex unit test was failing with 1.6rc1). If dict substitute is unsupported (it does seem to be undocumented) then it should raise an error, though not sure if there's another way to do this kind of variable substitution consistently.

**Comment by cmbant** [comment]
Incidentally, respite being rc, 1.6rc1 seems to be pulled by default from Travis python 3.6 ubuntu build, e..g see https://travis-ci.org/github/cmbant/CAMB/jobs/687446750

**Comment by smichr** [comment]
See [here](https://github.com/sympy/sympy/issues/6257#issuecomment-364794533) and containing issue.

**Comment by smichr** [comment]
But this is interesting. I am noting (and have been frustrated in #19261) by similar behavior that I was not able to track down. But the replacement order here should be `x` then `y` according to the docstring. And if so, then something about the `default_sort_key`, I am guessing, has changed since op count and nodes for tow symbols is the same so the replacements should be sorted according to the `default_sort_key` ordering of `x` and `y`.

**Comment by oscarbenjamin** [comment]
I'm not sure what should happen here but I think you should use `simultaneous=True` if you don't want the substitutions to interfere with each other:
```
(x*y).subs({x: x + 1, y: x}, simultaneous=True)
```
You can also pass the substitutions as an ordered iterable:
```
(x*y).subs({x: x + 1, y: x}.items())
```
I don't think we can guarantee to match dict insertion order while still supporting Python 3.5.

**Comment by oscarbenjamin** [comment]
It's hard to tell from the Travis script how exactly sympy is being installed. Maybe it's conda that picks up on prereleases from pypi. Maybe it depends on some of other conda commands that are being run.

**Comment by cmbant** [comment]
Thanks, yes, simultaneous=True is the workaround for the intended use (I think previously use with dict implied simultaneous?).

Note that symbols give a different result, which is very unintuitive
```

from sympy.abc import a, b
(a*b).subs({a: a + 1, b: a})
```
gives a(1+1).

**Comment by oscarbenjamin** [comment]
I don't know if passing a dict previously implied simultaneous=True but the docstring describes what is expected to happen:
https://github.com/sympy/sympy/blob/94fb720696f5f5d12bad8bc813699fd696afd2fb/sympy/core/basic.py#L800-L804
It seems that @smichr thinks that behaviour is not being respected in this case. I suspect that the specification in the docstring does not match what you would want (in general) anyway though.

**Comment by asmeurer** [comment]
I checked conda and it there are no conda packages built for 1.6 in the main or conda-forge channels. 

I also checked `pip install sympy` in a fresh environment and it installs 1.5.1. `pip install --pre sympy` does install 1.6rc1, as expected.

**Comment by asmeurer** [comment]
Maybe it has something to do with the fact that you use setup.py install. I wasn't able to run setup.py install on CAMB because I couldn't get all the Fortran stuff installed. I'm not clear why setup.py would install sympy pre-releases unless you have it configured to do so somehow.

**Comment by oscarbenjamin** [comment]
The subs issue here is a release blocker for 1.6.

@smichr is there anything that needs to be fixed for subs?

I've bisected the change to 132acff40e769f08dd834dd78f489ca56e2bed1d from #18043

To me the order is just undefined if a dict is passed in. It could be possible to use a topological sort but how do we even know what order the user wants?

Looking at the example in this issue
```python
(x*y).subs({x: x + 1, y: x})
```
if I saw this code then I would not know what the intention was. If the intention is dict order then we can't use that because it is not deterministic for all supported Python versions.

In sympy 1.5.1 which substitution happens depends on the name of the symbols:
```julia
In [1]: (x*y).subs({x:x+1, y:x})                                                                                                  
Out[1]: x⋅(x + 1)

In [2]: (x*y).subs({y:y+1, x:y})                                                                                                  
Out[2]: 
       2
(y + 1) 
```

**Comment by smichr** [comment]
> To me the order is just undefined if a dict is passed in

Yes, it's unordered so subs has to do something canonical with it. So the items are sorted: reverse `ordered` non-atoms followed by `ordered` atoms. The problem/difference is that the sorting of non-atoms -- in this case x(t) and y(t) -- if there is a tie of node length then perhaps they should be ordered.

```python
k = list(ordered(sequence))
# put highest node-count items first while retaining order
nd = defaultdict(list)
for i in k:
    nd[nodes(i)].append(i)
seq = []
for i in reversed(sorted(nd)):
    seq.extend(nd[i])
sequence = seq
```

**Comment by cmbant** [comment]
Regarding why 1.6 is installed, I'm not sure. This was not a conda build, the dependency of the package being installed is just sympy>=1; however the install log gives
```

Searching for sympy>=1.0
Reading https://pypi.org/simple/sympy/
Downloading https://files.pythonhosted.org/packages/4a/72/ad5bdd0a3c18e9bfc0ed532d0749729bb4b2813251c04d6b737df13e90c4/sympy-1.6rc1-py3-none-any.whl#sha256=d6d0c4990b9789699264e2db70979666289bf48c13fccc6e30aad3063374864d
Best match: sympy 1.6rc1
```
My unchanged travis tests on python 3.6 ubuntu (and only that one) just suddenly stopped working around the 14th May due to it changing to pulling 1.6rc1.

**Comment by asmeurer** [comment]
I'm not clear why setup.py install would install a pre-release for SymPy. These sorts of issues are why I have avoided uploading pre-releases to PyPI in the past.

**Comment by oscarbenjamin** [comment]
I'm reopening this because it is only fixed in the 1.6 branch and not master

**Comment by oscarbenjamin** [comment]
This was fixed by c1fed2c8899 in #19261
